specifyweb/stored_queries/batch_edit.py

```diff
@@ -0,0 +1,1184 @@
+    if field.is_relationship:
+        return f"{fieldspec.table.name} ({'formatted' if field.type.endswith('to-one') else 'aggregatd'})"
+    return f"{fieldspec.table.name} {field.name}"
+
+
+def run_batch_edit(collection, user, spquery, agent):
+    props = BatchEditProps(
+        collection=collection,
+        user=user,
+        contexttableid=int(spquery["contexttableid"]),
+        captions=spquery.get("captions", None),
+        limit=spquery.get("limit", 0),
+        recordsetid=spquery.get("recordsetid", None),
+        fields=fields_from_json(spquery["fields"]),
+        session_maker=models.session_context,
+        omit_relationships=True
+    )
+    (headers, rows, packs, json_upload_plan, visual_order) = run_batch_edit_query(props)
+    mapped_raws = [
+        [*row, json.dumps({"batch_edit": pack})] for (row, pack) in zip(rows, packs)
+    ]
+    # Skipping empty because we can have a funny case where all the query fields don't contain any data
+    regularized_rows = regularize_rows(len(headers), mapped_raws, skip_empty=False)
+    return make_dataset(
+        user=user,
+        collection=collection,
+        name=spquery["name"],
+        headers=headers,
+        regularized_rows=regularized_rows,
+        agent=agent,
+        json_upload_plan=json_upload_plan,
+        visual_order=visual_order,
+    )
+
+
+# @transaction.atomic <--- we DONT do this because the query logic could take up possibly multiple minutes
+class BatchEditProps(TypedDict):
+    collection: Any
+    user: Any
+    contexttableid: int
+    captions: Any
+    limit: Optional[int]
+    recordsetid: Optional[int]
+    session_maker: Any
+    fields: List[QueryField]
+    omit_relationships: Optional[bool]
+
+
+def run_batch_edit_query(props: BatchEditProps):
+
+    offset = 0
+    tableid = int(props["contexttableid"])
+    captions = props["captions"]
+    limit = props["limit"]
+
+    recordsetid = props["recordsetid"]
+    fields = props["fields"]
+
+    visible_fields = [field for field in fields if field.display]
+
+    assert captions is None or (
+        len(visible_fields) == len(captions)
+    ), "Got misaligned captions!"
+
+    localization_dump: Dict[str, str] = (
+        {
+            # we cannot use numbers since they can very off
+            field.fieldspec.to_stringid(): caption
+            for field, caption in zip(visible_fields, captions)
+        }
+        if captions is not None
+        else {}
+    )
+
+    row_plan = RowPlanMap.get_row_plan(visible_fields)
+
+    indexed, query_fields = row_plan.index_plan()
+    # we don't really care about these fields, since we'have already done the numbering (and it won't break with
+    # more fields). We also don't caree about their sort, since their sort is guaranteed to be after ours
+    query_with_hidden = [
+        *query_fields,
+        *[field for field in fields if not field.display],
+    ]
+
+    with props["session_maker"]() as session:
+        rows = execute(
```

Comment:
I recently learned that this is not the most optimum way of fetching rows. We actually use a different way (even for exports) - which is to use yield_per
https://github.com/specify/specify7/blob/bad8e594f3395e4b656db00ccce03102c69a741c/specifyweb/stored_queries/execution.py#L221
This is because otherwise it fetches all at once, and dedicates memory for the entire set. 
See: https://docs.sqlalchemy.org/en/13/orm/query.html#sqlalchemy.orm.query.Query.yield_per
Maybe refactor this usage to use yield_per?